src/evm/wasm.ts

```diff
@@ -0,0 +1,68 @@
+import fs from "fs";
+import path from "path";
+
+export class Wasm {
+    private readonly instance: WebAssembly.Instance;
+
+    private constructor(instance: WebAssembly.Instance) {
+        this.instance = instance;
+    }
+
+    static async init(wasmBytes: Uint8Array): Promise<Wasm> {
+        const mod = await WebAssembly.compile(wasmBytes);
+        const instance = await WebAssembly.instantiate(mod);
+        return new Wasm(instance);
+    }
+
+    static async load(wasmPathOrUrl: string): Promise<Wasm> {
+        try {
+            let wasmBytes: Uint8Array;
+
+            if (typeof window !== "undefined") {
+                const response = await fetch(wasmPathOrUrl);
+                if (!response.ok) {
+                    throw new Error(
+                        `Failed to fetch WASM file: ${response.statusText}`
+                    );
+                }
+                const buffer = await response.arrayBuffer();
+                wasmBytes = new Uint8Array(buffer);
+            } else {
+                const resolvedPath = path.resolve(wasmPathOrUrl);
+                wasmBytes = fs.readFileSync(resolvedPath);
```

Comment:
I know you check the environment and this conde won't run in browsers, but it could still be problematic for the bundler that's building the UI application - there are probably some npm packages that make this polymorphic and bundler safe